### PR TITLE
Update AssemblyScript and other examples

### PR DESCRIPTION
**Pull request by MaxGraey**
With this changes + latest AssemblyScript (old one was aclually prototype) binary size reduced from  **9.33 KB** to **1 KB**.

**NOTE** Not tested in runtime so may contain bugs

### CODE REVIEW

File: assemblyscript/mandelbrot.ts [hunk 1]
```diff
@@ -25,27 +28,35 @@ function iterateEquation(x0: f64, y0: f64, maxiterations: i32): i32 {
...
     }
   }
 }
 
 export function getData(): Uint8Array {
   return data;
+}
+
+export function getDataBuffer(): ArrayBuffer {
+  return <ArrayBuffer>(data.buffer);
```
Comment (MaxGraey) [inline assemblyscript/mandelbrot.ts]: This may be need addisional offset:
```ts
return <ArrayBuffer>(<usize>data.buffer + 8);
```

**Comment by ColinEberhardt** [comment]
Thanks for the PR, very much appreciated. Just reviewing and testing now.

**Comment by MaxGraey** [comment]
So main goal of this PR see how wasm, browsers and compilers improved for last 9 month comparing to javascript.

File: js/mandelbrot.js [hunk 2]
```diff
@@ -24,22 +24,25 @@ const iterateEquation = (x0, y0, maxiterations) => {
   return iterations;
 }
 
-const scale = (domainStart, domainLength, screenLength, step) =>
-  domainStart + domainLength * ((step - screenLength) / screenLength);
+function scale(domainStart, domainLength, invScreenLength, step) {
+  domainStart + domainLength * (step * invScreenLength - 1);
```
Comment (ColinEberhardt) [inline js/mandelbrot.js]: you're missing a `return` statement here. Any reason for changing from function expressions to declarations?

Comment (MaxGraey) [reply js/mandelbrot.js]: Yah. Good catch!

Comment (MaxGraey) [reply js/mandelbrot.js]: Fixed

File: js/mandelbrot.js [hunk 3]
```diff
@@ -1,17 +1,17 @@
-const WIDTH = 1200, HEIGHT = 800;
+const WIDTH  = 1200;
```
Comment (ColinEberhardt) [inline js/mandelbrot.js:1]: at runtime the JS implementation is rendering a different image (compared to the C / WASM implementation) - looks like the colour scale has changed.

Comment (MaxGraey) [reply js/mandelbrot.js:1]: Fixed

File: js/mandelbrot.js [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
-  iteration = ((iteration * scale) + offset) % 1024;
+function colour(iteration, offset, scale) {
+  iteration = (iteration * scale + offset) & 1023;
   if (iteration < 256) {
     return iteration;
   } else if (iteration < 512) {
-    return 255 - (iteration - 255);
-  } else {
-    return 0;
+    return 255 - ((iteration & 255) - 255);
```
Comment (ColinEberhardt) [inline js/mandelbrot.js]: Ah - with respect to https://github.com/ColinEberhardt/wasm-mandelbrot/pull/3/files#r200177666 - the color scale has changed here

Comment (MaxGraey) [reply js/mandelbrot.js]: Yap, already fixed)

File: wasm/mandelbrot.c [hunk 5]
```diff
@@ -26,21 +25,23 @@ int iterateEquation(double x0, double y0, int maxiterations) {
   return iterations;
 }
 
-double scale(double domainStart, double domainLength, int screenLength, int step) {
-  return domainStart + domainLength * ((double)(step - screenLength) / (double)screenLength);
+inline double scale(double domainStart, double domainLength, double invScreenLength, int step) {
```
Comment (ColinEberhardt) [inline wasm/mandelbrot.c]: changing this function to inline causes the generated wasm binary to expect this function to be imported:

```
(import "env" "scale" (func $scale (param f64 f64 f64 i32) (result f64)))
```

Was that intended?

Comment (MaxGraey) [reply wasm/mandelbrot.c]: Nope. Hmm, that's strange

Comment (MaxGraey) [reply wasm/mandelbrot.c]: Ok, I'll remove this. Anyway GCC/Clang should perfectly inline this.

File: assemblyscript/mandelbrot.ts [hunk 6]
```diff
@@ -25,27 +28,35 @@ function iterateEquation(x0: f64, y0: f64, maxiterations: i32): i32 {
   return iterations;
 }
 
-function scale(domainStart: f64, domainLength: f64, screenLength: f64, step: f64): f64 {
-  return domainStart + domainLength * ((step - screenLength) / screenLength);
+@inline
```
Comment (ColinEberhardt) [inline assemblyscript/mandelbrot.ts:31]: When compiling this file with the TypeScript compiler it fails as follows:

~~~
error TS1206: Decorators are not valid here.
~~~

When I first looked at AssemblyScript I really liked the way you could compile the same codebase to either wasm or JS. I found this feature to be quite useful:

https://blog.scottlogic.com/2017/10/30/migrating-d3-force-layout-to-webassembly.html

Is this something you'd expect AssemblyScript to support

Comment (MaxGraey) [reply assemblyscript/mandelbrot.ts:31]: Yes. This quite tricky and may be @dcodeIO helps with this. I can't prepend `portable.js` which contain portable polyfills for AS when target is JS. So for compiling to JS I temprary comment `@inline` decorator and after remove `unchecked` wrappers from generated js source.

**Comment by ColinEberhardt** [comment]
Thanks for the PR - really appreciate this, although there are a few runtime errors that need to be resolved. The good news is that the AssemblyScript version worked first time!

**Comment by MaxGraey** [comment]
Cool! It would be just perfect to see new results)

**Comment by ColinEberhardt** [comment]
> It would be just perfect to see new results

Are you interested in performance specifically? i.e. the older version of ASC versus newer?

**Comment by MaxGraey** [comment]
Yes indeed. Older vs Newer as well=) But especially how far wasm from JS and  C baseline

**Comment by ColinEberhardt** [comment]
I'm struggling to get the emscripten / asm.js versions to build right now - I'm on a train (bad WiFi), so can't investigate fully. I'll give this a go tomorrow and see if I can get some performance measurements.

**Comment by MaxGraey** [comment]
Sure. Have a good trip!

**Comment by ColinEberhardt** [comment]
Had another look at this - unfortunately quite a bit has changed in the emscripten tooling, and I just don't have time to update the build scripts on this projects!

As a compromise, I'd be happy to merge a change that simply updates the AssemblyScript code to the latest version, without the optimisations you've made to the mandelbrot algorithm (this is the same across all implementations, so no need to really optimise it!)

**Comment by MaxGraey** [comment]
Hmm, that's sad but if you update AS and some tweaks for all JS sides. this make me happy 👍

**Comment by MaxGraey** [comment]
Hi, @ColinEberhardt. Will you plan to merge this PR or partially update AS example?

**Comment by ColinEberhardt** [comment]
Thanks for the reminder @MaxGraey - I've updated the AS example using your code https://github.com/ColinEberhardt/wasm-mandelbrot/commit/a9e6666da60be88fdff14422f1def46f1badd769

I did a quick performance comparison:

- previous AS version 584ms
- new AS version 561ms

(averaged over x10 iterations).

It looks faster, but I didn't measure the standard deviation, so could be isn't entirely conclusive!

**Comment by MaxGraey** [comment]
Thanks!

**Pull request closed without merging**